### PR TITLE
WIP trying to add intrinsic filter

### PR DESCRIPTION
**Pull request by yairchn**
This is not complete and not running yet - I would appreciate comments on fixing it

### CODE REVIEW

File: test/Atmos/EDMF/edmf_model.jl [hunk 1]
```diff
@@ -413,6 +413,25 @@ function EDMF(
...
+    Q::FT
+    "filter edmf environment TKE"
+    TKE::FT
+    "filter edmf environment cv_θ"
+    cv_θ::FT
+    "filter edmf environment cv_Q"
+    cv_Q::FT
+    "filter edmf environment cv_θQ"
+    cv_θQ::FT
+end
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/edmf_model.jl:433]: These properties need to match the `target` properties above. And `::FT` should probably be `::B` (they're Boolean, right?)

File: test/Atmos/EDMF/ekman_layer.jl [hunk 2]
```diff
@@ -187,6 +190,16 @@ function main(::Type{FT}, cl_args) where {FT}
 
     dgn_config = config_diagnostics(driver_config)
 
+    edmf_filter_targets = (a = true,
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/ekman_layer.jl:193]: These will need to match what we're using above, `filter_ρa`, `filter_ρaw` etc.

File: test/Atmos/EDMF/edmf_model.jl (hunk 1 above)
Comment (yairchn) [reply test/Atmos/EDMF/edmf_model.jl:433]: you mean AbstractFloat should be replaced by 'Bool' or something similar ?

Comment (charleskawczynski) [reply test/Atmos/EDMF/edmf_model.jl:433]: I mean it should look like:
```julia
Base.@kwdef struct IntrinsicFilter{B <: Bool} <: AbstractFilterTarget
    "filter edmf updraft a"
    filter_ρa::B
    "filter edmf updraft w"
    filter_ρaw::B
...
```

Comment (yairchn) [reply test/Atmos/EDMF/edmf_model.jl]: I changed the names as well the 'filter_' was not adding any information

File: test/Atmos/EDMF/edmf_kernels.jl [hunk 3]
```diff
@@ -361,6 +367,59 @@ function primitive_to_prognostic!(
 
 end
 
+function compute_filter_argument!(
+    target::target,
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/edmf_kernels.jl:371]:
```suggestion
    target::MomentumPerturbation,
```

File: test/Atmos/EDMF/ekman_layer.jl [hunk 4]
```diff
@@ -187,6 +190,16 @@ function main(::Type{FT}, cl_args) where {FT}
 
     dgn_config = config_diagnostics(driver_config)
 
+    edmf_filter_targets = (ρa = true,
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/ekman_layer.jl:193]:
```suggestion
    edmf_filter_targets = (;ρa = true,
```

File: test/Atmos/EDMF/edmf_model.jl [hunk 5]
```diff
@@ -413,6 +413,34 @@ function EDMF(
...
+    "filter edmf updraft Q"
+    ρaq_tot::B
+    "filter edmf environment TKE"
+    ρatke::B
+    "filter edmf environment cv_θ"
+    ρaθ_liq_cv::B
+    "filter edmf environment cv_Q"
+    ρaq_tot_cv::B
+    "filter edmf environment cv_θQ"
+    ρaθ_liq_q_tot_cv::B
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/edmf_model.jl:441]:
```suggestion
    "filter edmf updraft a"
    ρa::B = false
    "filter edmf updraft w"
    ρaw::B = false
    "filter edmf updraft θ"
    ρaθ_liq::B = false
    "filter edmf updraft Q"
    ρaq_tot::B = false
    "filter edmf environment TKE"
    ρatke::B = false
    "filter edmf environment cv_θ"
    ρaθ_liq_cv::B = false
    "filter edmf environment cv_Q"
    ρaq_tot_cv::B = false
    "filter edmf environment cv_θQ"
    ρaθ_liq_q_tot_cv::B = false
```

File: test/Atmos/EDMF/edmf_model.jl [hunk 6]
```diff
@@ -413,6 +413,34 @@ function EDMF(
...
+Base.@kwdef struct IntrinsicFilter{B <: Bool} <: AbstractFilterTarget
+# Base.@kwdef struct IntrinsicFilter{Bool,ρa,
+#                                         ρaw,
+#                                         ρaθ_liq,
+#                                         ρaq_tot,
+#                                         ρatke,
+#                                         ρaθ_liq_cv,
+#                                         ρaq_tot_cv,
+#                                         ρaθ_liq_q_tot_cv,
+#                                         } <: AbstractFilterTarget
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/edmf_model.jl:425]:
```suggestion
```

File: test/Atmos/EDMF/ekman_layer.jl [hunk 7]
```diff
@@ -187,6 +190,16 @@ function main(::Type{FT}, cl_args) where {FT}
...
+    edmf_filter_targets = (;ρa = true,
+                           ρaw = true,
+                           ρaθ_liq = true,
+                           ρaq_tot = true,
+                           ρatke = true,
+                           ρaθ_liq_cv = true,
+                           ρaq_tot_cv = true,
+                           ρaθ_liq_q_tot_cv = true)
+
+    IntrinsicFilter(model; target = edmf_filter_targets)
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/ekman_layer.jl:202]:
```suggestion
    edmf_intrinsic_filter = IntrinsicFilter(;ρa = true,
                           ρaw = true,
                           ρaθ_liq = true,
                           ρaq_tot = true,
                           ρatke = true,
                           ρaθ_liq_cv = true,
                           ρaq_tot_cv = true,
                           ρaθ_liq_q_tot_cv = true)
```

File: test/Atmos/EDMF/edmf_model.jl [hunk 8]
```diff
@@ -413,6 +413,24 @@ function EDMF(
     return EDMF{FT, N_up, typeof.(args)...}(args...)
 end
 
+Base.@kwdef struct IntrinsicEDMFFilter{B <: Bool} <: AbstractFilterTarget
+    "filter edmf updraft a"
+    ρa::B = false
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/edmf_model.jl:418]:
```suggestion
Base.@kwdef struct IntrinsicEDMFFilter{A, B <: Bool} <: AbstractFilterTarget
    "The AtmosModel"
    atmos::A
    "filter edmf updraft a"
    ρa::B = false
```

File: test/Atmos/EDMF/ekman_layer.jl [hunk 9]
```diff
@@ -187,11 +190,21 @@ function main(::Type{FT}, cl_args) where {FT}
 
     dgn_config = config_diagnostics(driver_config)
 
+    edmf_filter_targets = IntrinsicEDMFFilter(;ρa = true,
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/ekman_layer.jl:193]:
```suggestion
    edmf_filter = IntrinsicEDMFFilter(solver_config.bl; ρa = true,
```

File: test/Atmos/EDMF/ekman_layer.jl [hunk 10]
```diff
@@ -187,11 +190,21 @@ function main(::Type{FT}, cl_args) where {FT}
...
+                           ρaq_tot_cv = true,
+                           ρaθ_liq_q_tot_cv = true)
+
     # boyd vandeven filter
     cb_boyd = GenericCallbacks.EveryXSimulationSteps(1) do
         Filters.apply!(
             solver_config.Q,
-            ("energy.ρe",),
+            # IntrinsicEDMFFilter(model = driver_config.bl; edmf_filter_targets...),
+            IntrinsicEDMFFilter(edmf_filter_targets...),
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/ekman_layer.jl:207]:
```suggestion
            edmf_filter,
```

File: test/Atmos/EDMF/edmf_kernels.jl [hunk 11]
```diff
@@ -361,6 +367,59 @@ function primitive_to_prognostic!(
 
 end
 
+function compute_filter_argument!(
+    target::IntrinsicEDMFFilter,
+    filter_state::Vars,
+    state::Vars,
+    aux::Vars,
+)
+    gm = state
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/edmf_kernels.jl:376]:
```suggestion
    N_up = n_updrafts(turbconv_model(target.atmos))
    gm = state
```

File: test/Atmos/EDMF/edmf_kernels.jl [hunk 12]
```diff
@@ -361,6 +367,59 @@ function primitive_to_prognostic!(
...
+    up = state.turbconv.updraft
+    en = state.turbconv.environment
+    fs_up = filter_state.turbconv.updraft
+    fs_en = filter_state.turbconv.environment
+    ρ_inv = 1 / gm.ρ
+
+    ft_en = target.environment
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/edmf_kernels.jl:383]:
```suggestion
    ft_en = filter_state.turbconv.environment
```
+ similar changes below

File: test/Atmos/EDMF/edmf_kernels.jl [hunk 13]
```diff
@@ -77,6 +77,9 @@ function vars_state(::Environment, ::Primitive, FT)
     @vars(atke::FT, aθ_liq_cv::FT, aq_tot_cv::FT, aθ_liq_q_tot_cv::FT,)
 end
 
+vars_state_filtered(target::IntrinsicEDMFFilter,FT) =
+    vars_state_filtered(target.atmos, Primitive(), FT)
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/edmf_kernels.jl:81]:
```suggestion
    vars_state(target.atmos, Primitive(), FT)
```

**Pull request closed without merging**